rfc/rfc-78/rfc-78.md

```diff
@@ -0,0 +1,339 @@
+lf1 and lf2 has delta commit time as t3 (which matches the base file's instant time) and its completion time is t10. 
+
+lets trigger an upgrade and add 2 new files. 
+
+fs3(t3):
+lf1_(3_10) lf2_(3_10) [upgrade] lf3_15_20, lf4_40_100.
+
+With this layout, 1.x reader should be able to read this file slice correctly w/o any issues. So, we should not require a full compaction during upgrade. But need to consult w/ authors to ensure we are not missing anything.
+
+## 1.0 ➝ 0.16.0 downgrade
+Any new features that was introduced in 1.x may not work after downgrade.
+Here are the ones that might see behavior change: 
+- If deletion vector is enabled, after downgrade we might fallback to key based merges, giving up on the performance optimization.
+- If partial updates are enabled, after downgrade we might fallback to key based merges, giving up on the performance optimization.
+- Any functional indexes built in 1.x might be deleted during downgrade. 
+- Any secondary indexes built in 1.x might be deleted during downgrade. 
+- Completion time based log file ordering within a file slice may not be honored.
+
+### Users will have to use hudi-cli to do the downgrade. Here are the steps we might need to do during downgrade
+- Similar to our upgrade, lets trigger full compaction to latest file slice so that all latest file slices only has base files. We do not want to have any file slices w/ a mix of 1.x log files and 0.x log files. 
+We could afford to take some downtime during downgrade, but upgrade path should be kept to as minimum as possible. So, during upgrade, we are not enforcing this.   
+- Rewrite LSM based timeline to archived timeline. we have to deduce writer properties and introduce boundaries based on that (i.e. until which commit to archive).
+- We have two options wrt handling active timeline instants(for eg, pending clustering instants):
+   A: No changes to active instants. In order to support 1.x tables w/ 0.16.0 reader, already these changes might have been ported to 0.16.0 reader. But we might have to ensure downgrade from 0.16.0 to any older version of hudi does take care of rewriting the active timeline instants.
+   B: While downgrading from 1.x to 0.16.0, let’s rewrite the active timeline instants too. So that after downgrade, we can't differentiate whether table was natively created in 0.16.0 or was it upgraded to 1.x and then later downgraded. The con in Option A is not an issue here since we take care of active timeline instants during 1.x to 0.16.0 downgrade only. So, downgrade from 0.16.0 to any lower hudi version does not need to do any special handling if table was native to 0.16.0 or was it downgraded from 1.x to 0.16.0
+   Our proposal is to go with Option B to ensure 0.16.0 table will be intact and not have any residues from 1.x writes. While the rewrite of active instants are happening, we could see some read failures. But 
+   we do not want to have a 0.16.0 (post upgrade) table to deal w/ 1.x timeline intricacies. 
```

Comment:
An optimization we thought that could help us in the downgrade flow or post downgrade flow: 
say we introduce a table property called "table.version.lineage", whose value is a comma separated list of all diff table versions this table has come through. 

for eg, "7,8". this means that, table was created w/ version 7 and then upgraded to 8. 

"7,8,7" means, table was with 7 and then upgraded to 8 and then downgraded to 7 again. 

so, by interception this config value we can deterministically say if table has been upgraded to 1.x before or not. If yes, we could completely avoid downgrading a 0.16.0 table further down (say 0.15.0.). 

If not, we should be allowing downgrade from 0.16.0 to 0.15.0. Reason is: say someone upgraded from 0.15.0 to 0.16.0 and due to some issues, they want to downgrade to 0.15.0. We should definitely allow them to downgrade. 

only sequence we want to add constraints is: after upgrading to 1.0.x, and downgraded to 0.16.0. After sometime, if they wish to downgrade to 0.15.0, we should avoid the downgrade.

Reply:
I don't prefer introducing these properties into table props. This is something the operator has to be aware of.

Reply:
noted. but do you have good recommendation on how we can guard further downgrades then?

Case 1: 
table created with 0.15.0 and then upgraded to 0.16.0. 
user runs into some issue. just executes the downgrade script and it works. 

Case2: 
table created with 0.15.0 and then upgraded to 0.16.0 and then upgraded to 1.x 
Lot of commits succeeded creating file slices completely in 1.x. 
user runs into some issue. Downgrades from 1.x to 0.16.0 will trigger full compaction. But there could be older file slices which were completely written in 1.x which we are not going to do anything as of current RFC design. 
And later user downgrades to 0.15.0. Downgrade succeeds. 
and now when user tries time travel or incr query, it might break or give unexpected results since log files are written in 1.x. 

So, if not for above proposal, only other option I can think of is: we can intercept the presence of log files in 1.x while downgrading from 0.16.x to 0.15.x and throw exception claiming that we can't downgrade. i.e. atleast throw clear error message.